Moved utils from explore.py to utils.py and moved one-hot encoding authors to a separate utility method

anton164 opened the pull request:
The changes are live here: [http://35.225.252.74/](http://35.225.252.74/)

**Update:** 
- Small experiment for vectorizing text using spacy
- Refactoring - splitting up into more utility methods and files. I will do another refactoring to make it more consistent with @janhenrikbern 's stuff after his PR is merged.

___

**Update:** Added language detection using pycld2
<img width="792" alt="Screenshot 2020-10-15 at 14 36 32" src="https://user-images.githubusercontent.com/1349225/96172217-1447f000-0ef4-11eb-9271-f3e9926f47c7.png">

___


This PR simplifies `explore.py` by pulling out a lot of the code into utility methods in `utils.py`. This should make it easier to re-use it in separate "notebooks".

I discovered that IP addresses for the compute engines are dynamic, so I've added a new step in the Github workflow that prints the IP address (you can also find the IP address in the gcp dashboard, but make sure to use `http://` not `https://`):
<img width="898" alt="Screenshot 2020-10-14 at 14 17 33" src="https://user-images.githubusercontent.com/1349225/96029021-0af15180-0e28-11eb-80f3-d9563d33333b.png">

anton164 commented on explore.py:
```diff
@@ -1,110 +1,61 @@
+    **Warning:** This is slow on large datasets
+    """
+)
+if st.button("Run one-hot encoding"):
```
`st.button` is a pretty nice utility for making expensive computations "optional"

anton164 commented on utils.py:
```diff
@@ -0,0 +1,143 @@
+
+# Due to https://github.com/mikemccand/chromium-compact-language-detector/issues/22
+# cld2 can't handle unprintable characters
+def strip_unprintable(s):
```
This adds a bit of runtime to `load_dataset` and is strictly not necessary for all abstracts. We could instead catch exceptions in `detect_language` and strip unprintable characters if it fails.

Perhaps the best approach is to eventually modify our `load_dataset` method to detect language upfront when loading the dataset and simply drop all non-english text.

janhenrikbern commented:
this looks good. We can optimize the loading sequence in a future PR after we finish the exploration stage.

I would suggest we create a `utils/` and then separate by functionality, e.g., `utils/data/loader.py'. While this will create a lot of files it will be easy to contribute to `util` with a minimum amount of merge conflicts. 

This PR gets a 👍 from me

elijah-rou commented on utils.py:
```diff
@@ -0,0 +1,143 @@
+
+
+@st.cache(suppress_st_warning=True)
+def load_dataset(dataset_filename, limit):
```
Do we have a custom method just for the progress bar? What's the difference over pd.read_json?

anton164 replied:
This method reads each json line in the file and creates a dataframe. It is different from `pd.read_json`, since it handles nested json fields/array fields, deletes some irrelevant fields & strips irrelevant characters. 

Using `pd.read_json` on the whole json file would be slower and more memory-intensive since we would need to read the whole file into memory and then handle each column after.

anton164 commented:
@janhenrikbern / @elijah-rou I've updated my PR and moved the utils into `data_tools.py`

The last commit is basically just moving the methods

You can see it live here: http://35.225.252.74/

Good to go?

anton164 commented:
I'm rebasing & merging to kill this long-living branch, let me know if you have any comments post-merge!
